appinventor/components/src/com/google/appinventor/components/runtime/NxtDirectCommands.java

```diff
@@ -142,56 +151,98 @@ public void PlayTone(int frequencyHz, int durationMs) {
   }
 
   @SimpleFunction(description = "Sets the output state of a motor on the robot.")
-  public void SetOutputState(String motorPortLetter, int power, int mode, int regulationMode,
-      int turnRatio, int runState, long tachoLimit) {
+  public void SetOutputState(
+    @Options(NxtMotorPort.class) String motorPortLetter,
+    int power,
+    @Options(NxtMotorMode.class) int mode,
+    @Options(NxtRegulationMode.class) int regulationMode,
+    int turnRatio,
+    @Options(NxtRunState.class) int runState,
+    long tachoLimit
+  ) {
     String functionName = "SetOutputState";
     if (!checkBluetooth(functionName)) {
       return;
     }
 
-    int port;
-    try {
-      port = convertMotorPortLetterToNumber(motorPortLetter);
-    } catch (IllegalArgumentException e) {
+    // Make sure motorPortLetter is a valid NxtMotorPort.
+    NxtMotorPort port = NxtMotorPort.fromUnderlyingValue(motorPortLetter);
+    if (port == null) {
       form.dispatchErrorOccurredEvent(this, functionName,
           ErrorMessages.ERROR_NXT_INVALID_MOTOR_PORT, motorPortLetter);
       return;
     }
 
-    setOutputState(functionName, port, power, mode,
-        regulationMode, sanitizeTurnRatio(turnRatio), runState, tachoLimit);
+    // Make sure mode is a valid NxtMotorMode.
+    NxtMotorMode motorMode = NxtMotorMode.fromUnderlyingValue(mode);
+    if (motorMode == null) {
+      return;
```

Comment:
I think if these are invalid we should report it. Previously, we just wrote through regardless of whether the value was correct, so returning here is a change in behavior.